### PR TITLE
Home work 2 function

### PR DESCRIPTION
**Pull request by FlyOn21**

### CODE REVIEW

File: hw_2.py [hunk 1]
```diff
@@ -1,15 +1,105 @@
+import math
+import random
+import sys
+
+
 # 1) Write a function that emulates the game "rock, scissors, paper"
 # At the entrance, your function accepts your version printed from the console, the computer makes a decision randomly.
+class Game_Rock_Scissors_Paper():
```
Comment (katevoskresenska) [inline hw_2.py:8]: Class names should use CamelCase convention

File: hw_2.py [hunk 2]
```diff
@@ -1,15 +1,105 @@
...
+        self.tuple_combination_win = (('rock', 'scissors'), ('scissors', 'paper'), ('paper', 'rock'))
+        self.tuple_combination_draw = (('rock', 'rock'), ('scissors', 'scissors'), ('paper', 'paper'))
+        self.dict_variations_input = {'1': 'rock', '2': 'scissors', '3': 'paper'}
+        self.mistake_counter = 0
+        self.replay_counter = 0
+
+    def main(self):
+        """Main function to start the game"""
+        people_choice = self.my_choice()
+        if people_choice is not False:
```
Comment (katevoskresenska) [inline hw_2.py:20]: if people_choice:

File: hw_2.py [hunk 3]
```diff
@@ -1,15 +1,105 @@
...
+            print(result)
+            self.play_again()
+        else:
+            print('May be you don\'t want play. By-by')
+            sys.exit()
+
+
+    def my_choice(self):
+        """Function implemented logic people choice"""
+        if self.check_counter_mistake() is not False:
```
Comment (katevoskresenska) [inline hw_2.py:32]: if self.check_counter_mistake():

File: hw_2.py [hunk 4]
```diff
@@ -49,4 +152,61 @@
 #     ["X", "X", "O"]
 # ]) ➞ "Draw"
 
+def main(tic_tac_toe):
+    """Function defines who winning in 3x3 matrix of a completed tic-tac-toe game"""
+    game_board = join_array(tic_tac_toe)
+    marks = ['X', 'O']
+    for mark in marks:
+        if win_check(game_board, mark) is True:
```
Comment (katevoskresenska) [inline hw_2.py:160]: if win_check(game_board, mark):

File: hw_2.py [hunk 5]
```diff
@@ -49,4 +152,61 @@
...
+        return True
+    elif mark == game_board[2] and mark == game_board[4] and mark == game_board[6]:
+        return True
+    elif mark == game_board[0] and mark == game_board[3] and mark == game_board[6]:
+        return True
+    elif mark == game_board[2] and mark == game_board[5] and mark == game_board[8]:
+        return True
+    elif mark == game_board[1] and mark == game_board[4] and mark == game_board[7]:
+        return True
+    return False
```
Comment (katevoskresenska) [inline hw_2.py:190]: I think usage of the logical operator can make your code more compact. 

    if mark == game_board[0] and mark == game_board[1] and mark == game_board[2] or \
            mark == game_board[3] and mark == game_board[4] and mark == game_board[5] or \
            mark == game_board[6] and mark == game_board[7] and mark == game_board[8] or \
            mark == game_board[0] and mark == game_board[4] and mark == game_board[8] or \
            mark == game_board[2] and mark == game_board[4] and mark == game_board[6] or \
            mark == game_board[0] and mark == game_board[3] and mark == game_board[6] or \
            mark == game_board[2] and mark == game_board[5] and mark == game_board[8] or \
            mark == game_board[1] and mark == game_board[4] and mark == game_board[7]:
        return True
    return False

File: hw_2.py [hunk 6]
```diff
@@ -49,4 +152,61 @@
...
+    result = encrypt_data()
+    print(result)
 
+    print('Task #4')
+    tic_tac_toe = ([
+    ["X", "O", "X"],
+    ["O", "X", "O"],
+    ["O", "X", "X"]])
+    result = main(tic_tac_toe)
+    print(result)
```
Comment (katevoskresenska) [inline hw_2.py:212]: Good job! :)

File: hw_2.py [hunk 7]
```diff
@@ -1,15 +1,105 @@
+import math
+import random
+import sys
+
+
 # 1) Write a function that emulates the game "rock, scissors, paper"
```
Comment (mmykhailenko) [inline hw_2.py:6]: General comment for Task1.

Try not to complicate the code. Here we no need a class at all, also we do not need so many objects like
self.tuple_combination_win, self.tuple_combination_draw, self.dict_variations_input and so much methods.  Will be enough one function.

File: hw_2.py [hunk 8]
```diff
@@ -1,15 +1,105 @@
...
 # and the average person uses 57 sheets per day.
-
 # Create a function that will receive a dictionary with two key/values:
 # "people" ⁠— Number of people in the household.
 # "tp" ⁠— Number of rolls.
 # Return a statement telling the user if they need to buy more TP!
+def toilet_paper_supply(dict_tp=None):
+    """Function determines if there is enough toilet paper for a family for 14 days"""
+    if dict_tp is None:
+        dict_tp = {'people': 3, 'tp': 8}
```
Comment (mmykhailenko) [inline hw_2.py:95]: better return the warning or print some message like ('Income data is empty')

File: hw_2.py [hunk 9]
```diff
@@ -26,7 +116,20 @@
 # encrypt("karaca") ➞ "0c0r0kaca"
 # encrypt("burak") ➞ "k0r3baca"
 # encrypt("alpaca") ➞ "0c0pl0aca"
-
+def encrypt_data(given_input='banana', encrypt_dict=None):
+    """Function implemented processing encrypt given input"""
+    if encrypt_dict is None:
+        encrypt_dict = {'a': '0', 'e': '1', 'i': '2', 'o': '2', 'u': '3'}
+    first_step = [char for char in reversed(given_input)]
```
Comment (mmykhailenko) [inline hw_2.py:123]: no need to do list comprehension here. You could iterate over reverse(given input)

File: hw_2.py [hunk 10]
```diff
@@ -26,7 +116,20 @@
...
+    second_step = []
+    for char in first_step:
+        if char in encrypt_dict.keys():
```
Comment (mmykhailenko) [inline hw_2.py:126]: encrypt_dict.keys() change to encrypt_dict

File: hw_2.py [hunk 11]
```diff
@@ -26,7 +116,20 @@
...
+            char = encrypt_dict[char]
+            second_step.append(char)
+        else:
+            second_step.append(char)
```
Comment (mmykhailenko) [inline hw_2.py:130]: second step.append(char) - duplicated code. Remove else

File: hw_2.py [hunk 12]
```diff
@@ -26,7 +116,20 @@
...
+def encrypt_data(given_input='banana', encrypt_dict=None):
+    """Function implemented processing encrypt given input"""
+    if encrypt_dict is None:
+        encrypt_dict = {'a': '0', 'e': '1', 'i': '2', 'o': '2', 'u': '3'}
+    first_step = [char for char in reversed(given_input)]
+    second_step = []
+    for char in first_step:
+        if char in encrypt_dict.keys():
+            char = encrypt_dict[char]
+            second_step.append(char)
```
Comment (mmykhailenko) [inline hw_2.py:128]: move this line left for a few spaces

File: hw_2.py [hunk 13]
```diff
@@ -49,4 +152,61 @@
 #     ["X", "X", "O"]
 # ]) ➞ "Draw"
 
+def main(tic_tac_toe):
+    """Function defines who winning in 3x3 matrix of a completed tic-tac-toe game"""
+    game_board = join_array(tic_tac_toe)
+    marks = ['X', 'O']
```
Comment (mmykhailenko) [inline hw_2.py:158]: use tuple here

File: hw_2.py [hunk 14]
```diff
@@ -49,4 +152,61 @@
...
+    for mark in marks:
+        if win_check(game_board, mark) is True:
+            return f'Win "{mark}"'
+    else:
```
Comment (mmykhailenko) [inline hw_2.py:162]: not sure why this else is here

**Review by mmykhailenko — APPROVED** [review]
pls add .idea/ forlder to .gitignor or just remove it from PR

**Pull request closed without merging**